pkg/manager/member/tikv_group_member_manager.go

```diff
@@ -92,3 +145,537 @@ func (tgm *tikvGroupMemberManager) registerTiKVGroup(tg *v1alpha1.TiKVGroup, tc
+		if status.LastHeartbeatTime.IsZero() {
+			if oldStatus, ok := previousStores[status.ID]; ok {
+				klog.V(4).Infof("the pod:%s's store LastHeartbeatTime is zero,so will keep in %v", status.PodName, oldStatus.LastHeartbeatTime)
+				status.LastHeartbeatTime = oldStatus.LastHeartbeatTime
+			}
+		}
+
+		oldStore, exist := previousStores[status.ID]
+
+		status.LastTransitionTime = metav1.Now()
+		if exist && status.State == oldStore.State {
+			status.LastTransitionTime = oldStore.LastTransitionTime
+		}
+
+		stores[status.ID] = *status
+	}
+
+	//this returns all tombstone stores
+	tombstoneStoresInfo, err := pdCli.GetTombStoneStores()
+	if err != nil {
+		tc.Status.TiKV.Synced = false
+		return err
+	}
+	for _, store := range tombstoneStoresInfo.Stores {
+		if store.Store != nil && !pattern.Match([]byte(store.Store.Address)) {
+			continue
+		}
+		status := getTiKVStore(store)
+		if status == nil {
+			continue
+		}
+		tombstoneStores[status.ID] = *status
+	}
+
+	tg.Status.Synced = true
+	tg.Status.Stores = stores
+	tg.Status.TombstoneStores = tombstoneStores
+	tg.Status.Image = ""
+	c := filterContainer(set, "tikv")
+	if c != nil {
+		tc.Status.TiKV.Image = c.Image
+	}
+	return nil
+}
+
+func (tgm *tikvGroupMemberManager) tikvGroupStatefulSetIsUpgrading(tg *v1alpha1.TiKVGroup, set *apps.StatefulSet) (bool, error) {
+	if statefulSetIsUpgrading(set) {
+		return true, nil
+	}
+	selector, err := label.NewGroup().Instance(tg.Name).TiKV().Selector()
+	if err != nil {
+		return false, err
+	}
+	tikvPods, err := tgm.podLister.Pods(set.Namespace).List(selector)
+	if err != nil {
+		return false, err
+	}
+	for _, pod := range tikvPods {
+		revisionHash, exist := pod.Labels[apps.ControllerRevisionHashLabelKey]
+		if !exist {
+			return false, nil
+		}
+		if revisionHash != tg.Status.StatefulSet.UpdateRevision {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// TODO: add unit test
+func newServiceForTiKVGroup(tg *v1alpha1.TiKVGroup, svcName string) *corev1.Service {
+	ns := tg.Namespace
+	tgName := tg.Name
+	tgLabel := label.NewGroup().TiKV().Instance(tgName)
+	svcLabel := label.NewGroup().TiKV().Instance(tgName).UsedByPeer()
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      svcName,
+			Namespace: ns,
+			Labels:    svcLabel.Labels(),
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "peer",
+					Port:       20160,
+					TargetPort: intstr.FromInt(20160),
+					Protocol:   corev1.ProtocolTCP,
+				},
+				{
+					Name:       "status",
+					Port:       20180,
+					TargetPort: intstr.FromInt(20180),
+					Protocol:   corev1.ProtocolTCP,
+				},
+			},
+			Selector:                 tgLabel.Labels(),
+			PublishNotReadyAddresses: true,
+			ClusterIP:                "None",
+		},
+	}
+	return svc
+}
+
+// TODO: add unit test
+func getTikVConfigMapForTiKVGroup(tg *v1alpha1.TiKVGroup, tc *v1alpha1.TidbCluster) (*corev1.ConfigMap, error) {
```

Comment:
what is the different between `getTikVConfigMapForTiKV`,maybe we can have a function &v1alpha1.TiKVConfig{} => *corev1.ConfigMap.Spec{}

Reply:
nice catch, updated.